### PR TITLE
fix sql compatible null handling config work with runtime.properties

### PR DESCRIPTION
**Pull request by clintropolis**
### Description
This terrible PR modifies `NullHandling` to no longer have a default value since it appears in practice to overwrite the instance injected by `NullHandlingModule` through `requestStaticInjection`, causing the `druid.generic.useDefaultValueForNull` config to not function when it is in a `runtime.properties` or similar file, and only works when specified as an argument. The injected instance was in fact being created, it was just being overwritten immediately by the default value that was being set for it in `NullHandling`.

Since there is no longer a  default value, I have added a method that is just awful, `NullHandling.initializeForTests()`, which any tests or benchmarks or anything that does not wire up injection **_must_** use to avoid null pointer exceptions. This is bad, but seemed better than any alternative I could come up with, because `NullHandling.replaceWithDefault` is called in a lot of 'hot' places, so introducing an `if` statement or method call overhead from using a `Supplier` or something like that didn't seem worth risking the overhead for the sake of easier to write tests.

<hr>

This PR has:
- [x] been self-reviewed.
- [x] been tested in a test Druid cluster.

<hr>

##### Key changed/added classes in this PR
 * `NullHandling`
 * `NullValueHandlingConfig`

### CODE REVIEW

**Comment by gianm** [comment]
@clintropolis The Travis failure looks legit, e.g.,

```java
[ERROR] org.apache.druid.server.lookup.jdbc.JdbcDataFetcherTest$FetchTest.testFetch(org.apache.druid.server.lookup.jdbc.JdbcDataFetcherTest$FetchTest)

[ERROR]   Run 1: JdbcDataFetcherTest$FetchTest.testFetch:111->assertMapLookup:176 » NullPointer

[ERROR]   Run 2: JdbcDataFetcherTest$FetchTest.testFetch:111->assertMapLookup:176 » NullPointer

[ERROR]   Run 3: JdbcDataFetcherTest$FetchTest.testFetch:111->assertMapLookup:176 » NullPointer

[ERROR]   Run 4: JdbcDataFetcherTest$FetchTest.testFetch:111->assertMapLookup:176 » NullPointer
```

File: core/src/main/java/org/apache/druid/common/config/NullValueHandlingConfig.java [hunk 1]
```diff
@@ -32,7 +33,7 @@
   public NullValueHandlingConfig(@JsonProperty("useDefaultValueForNull") Boolean useDefaultValuesForNull)
   {
     this.useDefaultValuesForNull = useDefaultValuesForNull == null
-                                   ? true
+                                   ? Boolean.valueOf(System.getProperty(NULL_HANDLING_CONFIG_STRING, "true"))
```
Comment (gianm) [inline core/src/main/java/org/apache/druid/common/config/NullValueHandlingConfig.java]: I don't think this is necessary. The normal Druid injectorator stuff (what is being activated by the `JsonConfigProvider.bind` line in NullHandlingModule) handles reading from system properties too.

File: core/src/main/java/org/apache/druid/common/config/NullHandling.java [hunk 2]
```diff
@@ -50,9 +49,17 @@
    * It does not take effect in all unit tests since we don't use Guice Injection.
    */
   @Inject
-  private static NullValueHandlingConfig INSTANCE = new NullValueHandlingConfig(
-      Boolean.valueOf(System.getProperty(NULL_HANDLING_CONFIG_STRING, "true"))
-  );
+  private static NullValueHandlingConfig INSTANCE;
```
Comment (gianm) [inline core/src/main/java/org/apache/druid/common/config/NullHandling.java]: I suggest making a private `getInstance()` method, using that everywhere, and having that method throw a nice exception that asks you to either call `NullHandling.initializeForTests()` or include `NullHandlingModule`. That way, future benchmark and test writers don't get confused by null pointer exceptions.

**Comment by clintropolis** [comment]
closing in favor of #8908

**Comment by clintropolis** [comment]
This PR also fixed up a non-conflicting merge issue between #8779 and #8871

**Pull request merged**